lib/utils/genmap/genmap.go

```diff
@@ -0,0 +1,317 @@
+// since the entry will eventually get cleaned up naturally when it hits MaxFailures.
+func (m *GenMap[K, V]) Terminate(key K) {
+	m.rw.RLock()
+	ent, ok := m.entries[key]
+	m.rw.RUnlock()
+	if !ok {
+		return
+	}
+
+	m.rw.Lock()
+	ent, ok = m.entries[key]
+	if ok {
+		delete(m.entries, key)
+	}
+	m.rw.Unlock()
+	if !ok {
+		return
+	}
+	ent.cancel()
+}
+
+// Get loads the value associated with the target key. If the key was not yet being generated by this
+// map instance, generation will be started. This method is non-blocking, except in the case where
+// the key being accessed is/was just added to the mapping.
+func (m *GenMap[K, V]) Get(ctx context.Context, key K) (val V, err error) {
+	ent, _ := m.ensureEntry(key)
+
+	// optimistically load the value pointer (this will always succeed unless the genmap just
+	// started tracking this value, so over time most calls should follow this path).
+	if vp := ent.ptr.Load(); vp != nil {
+		return vp.read()
+	}
+
+	// note that we aren't selecting on the per-entry context. each entry always generates at
+	// lease one value before exiting, so we only care about exiting early if the map as a whole
+	// is closed.
+	select {
+	case <-ent.init:
+		return ent.ptr.Load().read()
+	case <-m.ctx.Done():
+		return val, ErrGenMapClosed
+	case <-ctx.Done():
+		return val, ctx.Err()
+	}
+}
+
+// ensureEntry gets the entry for the target key if it exists, falling back to setting up a new one.
+func (m *GenMap[K, V]) ensureEntry(key K) (ent *entry[V], isNew bool) {
+	m.rw.RLock()
+	ent, ok := m.entries[key]
+	m.rw.RUnlock()
+	if ok {
+		return ent, false
+	}
+
+	// create an entry-level context that is a child of the
+	// main close context.
+	ctx, cancel := context.WithCancel(m.ctx)
+
+	newEnt := &entry[V]{
+		init:      make(chan struct{}),
+		wantRegen: make(chan struct{}, 1),
+		ctx:       ctx,
+		cancel:    cancel,
+	}
+
+	m.rw.Lock()
+	// check if another entry was set up concurrently
+	ent, ok = m.entries[key]
+	if !ok {
+		m.entries[key] = newEnt
+	}
+	m.rw.Unlock()
+	if ok {
+		cancel()
+		return ent, false
+	}
+
+	// start background opts for new entry
+	go m.generate(key, newEnt)
+
+	return newEnt, true
+}
+
+// generate is the background process that generates/regenerates the value associated with a given key.
+func (m *GenMap[K, V]) generate(key K, entry *entry[V]) {
+	defer func() {
+		m.rw.RLock()
+		current := m.entries[key]
+		m.rw.RUnlock()
+		if current != entry {
+			return
+		}
+
+		m.rw.Lock()
+		if m.entries[key] == entry {
+			delete(m.entries, key)
+		}
+		m.rw.Unlock()
+	}()
+
+	t := time.NewTimer(m.cfg.Jitter(m.cfg.RegenInterval))
+	defer t.Stop()
+
+	var init bool
+	var errs int
+	for {
+		// drain the reload channel since we're about to reload
+		select {
+		case <-entry.wantRegen:
+		default:
+		}
+
+		// reset the timer before loading to help compensate
+		// for slow generators.
+		t.Stop()
+		select {
+		case <-t.C:
+		default:
+		}
```

Comment:
This is racey, we might fail to stop a timer and then attempt to receive from the channel before the timer fills it. We should use the return value of `t.Stop()` but we can't reset and drain a timer without knowing if we already pulled from it elsewhere, so in this case we might need a flag and something like this.

```suggestion
		if !t.Stop() && !tDrain {
			<-t.C
		}
		tDrain = true
```